# <span style="color:darkblue"> Today: Pandas Merge and Concat Methods</span>

<font size="5"> 

- You can complement this lesson with this Datacamp course: https://app.datacamp.com/learn/courses/joining-data-with-pandas

- The ```DataFrame.merge()``` and ```DataFrame.concat()``` methods are two ways to combine different data frames together 

- **Merge:** combines data on common column values

- **Concat:** combines data based on column names

## <span style="color:darkblue"> 1. The merge method</span>

<font size="5"> 

- When you want to combine data objects based on one or more keys --> **combine rows that share data**

- When using ```merge```, we need to provide two required arguments:

    - The **left** ```DataFrame```
    
    - The **right** ```DataFrame```

- Merge is a flexible method and consequently it has many options

- There is one option that is particularly relevant to understand: ```how``` --> **specifies the type of merge we want to execute**

![rdb_us](images/rdb_us_congress.png)

### **Types of merge**

<font size="5"> 

- There are four types of merges:

    1. inner
    
    2. outer 

    3. left 

    4. right

![merges](images/merges.png)


## <span style="color:darkblue"> Sample of the U.S. Congress relational data </span>

![rdb_us](images/rdb_us_congress.png)

<font size="5"> 

- **Example 1:** we want to know what proportion of bills actions were proposed by members of the Senate and what percentage by members of the House of Representatives

    1. Check how many unique ```member_id``` there are in the **bills actions** table

    2. Check how many unique ```member_id``` there are in the **congress member** table

    3. What type of merge should we use?

In [1]:
import pandas as pd

members = pd.read_csv('data/us_congress_member.csv')
actions = pd.read_csv('data/bills_actions.csv')

<font size="5"> 

- Let's try the inner merge first. [Here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html) are the Pandas options:

    - **how:** {‘left’, ‘right’, ‘outer’, ‘inner’, ‘cross’}, **default** ‘inner’

    - **on:** label or list

    - **left_on** and **right_on** in case the variables have different numbers

In [6]:
pd.merge(actions, members, on='member_id')

congress  bill_number bill_type  \
0          116         1029         s   
1          116         1160         s   
2          116         1309         s   
3          116         1434         s   
4          116         1636         s   
...        ...          ...       ...   
3298       116          840        hr   
3299       116          840        hr   
3300       116          987        hr   
3301       116          986        hr   
3302       116            9        hr   

                                                 action  \
0     S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
1     S.Amdt.2659 Amendment SA 2659 proposed by Sena...   
2     S.Amdt.1275 Amendment SA 1275 proposed by Sena...   
3     S.Amdt.1269 Amendment SA 1269 proposed by Sena...   
4     S.Amdt.2707 Amendment SA 2707 proposed by Sena...   
...                                                 ...   
3298  POSTPONED PROCEEDINGS - At the conclusion of d...   
3299  H.Amdt.23 Amendment (A004) offered by Mr. Berg...   
3300  H.Amdt.222 Amendment (A003) offered by Mr. Wel...   
3301  H.Amdt.203 Amendment (A006) offered by Mr. Hol...   
3302  H.Amdt.183 Amendment (A015) offered by Mr. Boy...   

                                   main_action     object  member_id  \
0     senate amendment proposed (on the floor)  amendment      858.0   
1     senate amendment proposed (on the floor)  amendment      858.0   
2     senate amendment proposed (on the floor)  amendment      858.0   
3     senate amendment proposed (on the floor)  amendment      858.0   
4     senate amendment proposed (on the floor)  amendment      858.0   
...                                        ...        ...        ...   
3298             other house amendment actions  amendment      485.0   
3299                   house amendment offered  amendment      485.0   
3300                   house amendment offered  amendment      922.0   
3301                   house amendment offered  amendment      406.0   
3302                   house amendment offered  amendment      127.0   

             full_name  last_name    member_title           state  party_name  \
0      Mitch McConnell  McConnell         Senator        Kentucky  Republican   
1      Mitch McConnell  McConnell         Senator        Kentucky  Republican   
2      Mitch McConnell  McConnell         Senator        Kentucky  Republican   
3      Mitch McConnell  McConnell         Senator        Kentucky  Republican   
4      Mitch McConnell  McConnell         Senator        Kentucky  Republican   
...                ...        ...             ...             ...         ...   
3298      Jack Bergman    Bergman  Representative        Michigan  Republican   
3299      Jack Bergman    Bergman  Representative        Michigan  Republican   
3300       Peter Welch      Welch  Representative         Vermont  Democratic   
3301    George Holding    Holding  Representative  North Carolina  Republican   
3302  Brendan F. Boyle      Boyle  Representative    Pennsylvania  Democratic   

     chamber  
0     Senate  
1     Senate  
2     Senate  
3     Senate  
4     Senate  
...      ...  
3298   House  
3299   House  
3300   House  
3301   House  
3302   House  

[3303 rows x 13 columns]

<font size="5"> 

- Is something going to change if we use ```how='left'```?

In [7]:
pd.merge(actions, members, on='member_id', how='left')

congress  bill_number bill_type  \
0          116         1029         s   
1          116         1031         s   
2          116         1160         s   
3          116         1199         s   
4          116         1208         s   
...        ...          ...       ...   
3298       116            9        hr   
3299       116            9        hr   
3300       116            9        hr   
3301       116            9        hr   
3302       116            9        hr   

                                                 action  \
0     S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
1     S.Amdt.2698 Amendment SA 2698 proposed by Sena...   
2     S.Amdt.2659 Amendment SA 2659 proposed by Sena...   
3     Committee on Health, Education, Labor, and Pen...   
4     Committee on the Judiciary. Reported by Senato...   
...                                                 ...   
3298  H.Amdt.172 Amendment (A004) offered by Ms. Kus...   
3299  H.Amdt.171 Amendment (A003) offered by Ms. Hou...   
3300  H.Amdt.170 Amendment (A002) offered by Ms. Oma...   
3301  POSTPONED PROCEEDINGS - At the conclusion of d...   
3302  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   

                                   main_action       object  member_id  \
0     senate amendment proposed (on the floor)    amendment      858.0   
1     senate amendment proposed (on the floor)    amendment      675.0   
2     senate amendment proposed (on the floor)    amendment      858.0   
3        senate committee/subcommittee actions  senate bill     1561.0   
4        senate committee/subcommittee actions  senate bill     1580.0   
...                                        ...          ...        ...   
3298                   house amendment offered    amendment       36.0   
3299                   house amendment offered    amendment      186.0   
3300                   house amendment offered    amendment      477.0   
3301             other house amendment actions    amendment      393.0   
3302                   house amendment offered    amendment        6.0   

               full_name  last_name    member_title           state  \
0        Mitch McConnell  McConnell         Senator        Kentucky   
1            Josh Hawley     Hawley         Senator        Missouri   
2        Mitch McConnell  McConnell         Senator        Kentucky   
3        Lamar Alexander  Alexander         Senator       Tennessee   
4         Lindsey Graham     Graham         Senator  South Carolina   
...                  ...        ...             ...             ...   
3298       Ann M. Kuster     Kuster  Representative   New Hampshire   
3299    Chrissy Houlahan   Houlahan  Representative    Pennsylvania   
3300          Ilhan Omar       Omar  Representative       Minnesota   
3301  Frank, Jr. Pallone    Pallone  Representative      New Jersey   
3302   Adriano Espaillat  Espaillat  Representative        New York   

      party_name chamber  
0     Republican  Senate  
1     Republican  Senate  
2     Republican  Senate  
3     Republican  Senate  
4     Republican  Senate  
...          ...     ...  
3298  Democratic   House  
3299  Democratic   House  
3300  Democratic   House  
3301  Democratic   House  
3302  Democratic   House  

[3303 rows x 13 columns]

<font size="5"> 

No difference between the two. They both have 3303 rows

<font size="5"> 

- How about if we use ```how='right'```?

In [9]:
pd.merge(actions, members, on='member_id', how='right')

congress  bill_number bill_type  \
0          NaN          NaN       NaN   
1          NaN          NaN       NaN   
2        116.0          3.0        hr   
3        116.0       7617.0        hr   
4        116.0       3055.0        hr   
...        ...          ...       ...   
4728     116.0       3055.0        hr   
4729     116.0       3469.0        hr   
4730     116.0       4739.0        hr   
4731     116.0       4761.0        hr   
4732       NaN          NaN       NaN   

                                                 action  \
0                                                   NaN   
1                                                   NaN   
2     H.Amdt.719 Amendment (A009) offered by Ms. Fin...   
3     H.Amdt.870 Amendment (A012) offered by Ms. Fin...   
4     POSTPONED PROCEEDINGS - At the conclusion of d...   
...                                                 ...   
4728  H.Amdt.401 Amendment (A015) offered by Ms. Cla...   
4729  Ms. Clarke (NY) moved to suspend the rules and...   
4730  Ms. Clarke (NY) moved to suspend the rules and...   
4731  Ms. Clarke (NY) moved to suspend the rules and...   
4732                                                NaN   

                        main_action      object  member_id  \
0                               NaN         NaN        0.0   
1                               NaN         NaN        1.0   
2           house amendment offered   amendment        2.0   
3           house amendment offered   amendment        2.0   
4     other house amendment actions   amendment        3.0   
...                             ...         ...        ...   
4728        house amendment offered   amendment     1809.0   
4729            house floor actions  house bill     1809.0   
4730            house floor actions  house bill     1809.0   
4731            house floor actions  house bill     1809.0   
4732                            NaN         NaN     1810.0   

                     full_name   last_name    member_title     state  \
0           A. Donald McEachin    McEachin  Representative  Virginia   
1                 Aaron Schock      Schock  Representative  Illinois   
2              Abby Finkenauer  Finkenauer  Representative      Iowa   
3              Abby Finkenauer  Finkenauer  Representative      Iowa   
4     Abigail Davis Spanberger  Spanberger  Representative  Virginia   
...                        ...         ...             ...       ...   
4728          Yvette D. Clarke      Clarke  Representative  New York   
4729          Yvette D. Clarke      Clarke  Representative  New York   
4730          Yvette D. Clarke      Clarke  Representative  New York   
4731          Yvette D. Clarke      Clarke  Representative  New York   
4732               Zell Miller      Miller         Senator   Georgia   

      party_name chamber  
0     Democratic   House  
1     Republican   House  
2     Democratic   House  
3     Democratic   House  
4     Democratic   House  
...          ...     ...  
4728  Democratic   House  
4729  Democratic   House  
4730  Democratic   House  
4731  Democratic   House  
4732  Democratic  Senate  

[4733 rows x 13 columns]

<font size="5"> 

Clearly the right merge has more rows and there are missing for the actions without a member

<font size="5"> 

- Let's save the correct merged data into a new object and perform the calculation we need

- How can I calculate the proportion of legislative actions from Senators and House Representatives 

In [ ]:
merged = pd.merge(actions, members, on='member_id', how='left')

In [ ]:
sn = sum(merged['member_title']=='Senator')
rn = sum(merged['member_title']=='Representative')

In [ ]:
sn/(sn+rn)

0.25615006150061503

In [ ]:
rn/(sn+rn)

0.743849938499385

## <span style="color:darkblue"> 1.1. Merging on more than one key</span>

<font size="5"> 

- Instead of passing only a name for the ```on``` option, we pass a list

- **Example 2:** how many actions are associated with the subject ```'Congressional oversight'```?

![rdb_us](images/rdb_us_congress.png)

In [11]:
subjects = pd.read_csv('data/bills_subjects.csv')

In [12]:
mergeds = pd.merge(actions, subjects, on = ['congress', 'bill_number', 'bill_type'])

In [13]:
mergeds

congress  bill_number bill_type  \
0            116         1029         s   
1            116         1029         s   
2            116         1029         s   
3            116         1029         s   
4            116         1029         s   
...          ...          ...       ...   
153967       116            9        hr   
153968       116            9        hr   
153969       116            9        hr   
153970       116            9        hr   
153971       116            9        hr   

                                                   action  \
0       S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
1       S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
2       S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
3       S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
4       S.Amdt.1274 Amendment SA 1274 proposed by Sena...   
...                                                   ...   
153967  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   
153968  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   
153969  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   
153970  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   
153971  H.Amdt.169 Amendment (A001) offered by Mr. Esp...   

                                     main_action     object  member_id  \
0       senate amendment proposed (on the floor)  amendment      858.0   
1       senate amendment proposed (on the floor)  amendment      858.0   
2       senate amendment proposed (on the floor)  amendment      858.0   
3       senate amendment proposed (on the floor)  amendment      858.0   
4       senate amendment proposed (on the floor)  amendment      858.0   
...                                          ...        ...        ...   
153967                   house amendment offered  amendment        6.0   
153968                   house amendment offered  amendment        6.0   
153969                   house amendment offered  amendment        6.0   
153970                   house amendment offered  amendment        6.0   
153971                   house amendment offered  amendment        6.0   

                                             bill_subject  
0                       Criminal procedure and sentencing  
1                                  Evidence and witnesses  
2                   Judicial procedure and administration  
3                                                 Mammals  
4                                         Service animals  
...                                                   ...  
153967  Presidents and presidential powers, Vice Presi...  
153968                   Rural conditions and development  
153969              State and local government operations  
153970                 U.S. territories and protectorates  
153971                                 Wages and earnings  

[153972 rows x 8 columns]

In [15]:
sum(mergeds['bill_subject'] == 'Congressional oversight')

1869

## <span style="color:darkblue"> 2. The concat method</span>

<font size="5"> 

- With concatenation, the datasets are pulled together along an axis: column or row

![concat](images/concat.png)

<font size="5"> 

- **Example 3:** concatenate the bills actions tables with and without ```member_id```